node/node_handler.go

```diff
@@ -288,6 +286,43 @@ var BigMaxUint64 = new(big.Int).SetBytes([]byte{
 	255, 255, 255, 255, 255, 255, 255, 255,
 })
 
+// validateNewShardState_v2 validate whether the new shard state root matches
+func (node *Node) validateNewShardState_v2(block *types.Block, stakeInfo *map[common.Address]*structs.StakeInfo) error {
+	// Common case first – blocks without resharding proposal
+	header := block.Header()
+	if header.ShardStateHash == (common.Hash{}) {
+		// No new shard state was proposed
+		if core.IsEpochLastBlock(block) {
+			// TODO invoke view change
+			return errors.New("beacon leader did not propose resharding")
+		}
+		// We aren't expecting to reshard, so proceed to sign
+		return nil
+	}
+	shardState := &types.ShardState{}
+	err := rlp.DecodeBytes(header.ShardState, shardState)
+	if err != nil {
+		return err
+	}
+	proposed := *shardState
+	nextEpoch := new(big.Int).Add(block.Header().Epoch, common.Big1)
+	expected, err := core.CalculateNewShardState(
```

Comment:
The new shard state are not manually set now?

Reply:
It is still manually set, but abstract into an API, so we don't need to change this code later